include/caffe/util/device_alternate.hpp

```diff
@@ -86,7 +86,13 @@ const int CAFFE_CUDA_NUM_THREADS = 512;
 
 // CUDA: number of blocks for threads.
 inline int CAFFE_GET_BLOCKS(const int N) {
-  return (N + CAFFE_CUDA_NUM_THREADS - 1) / CAFFE_CUDA_NUM_THREADS;
+  cudaDeviceProp prop;
+  int device;
+  cudaGetDevice(&device);
+  CUDA_CHECK(cudaGetDeviceProperties(&prop, device));
+  int num_blocks = (N + CAFFE_CUDA_NUM_THREADS - 1) / CAFFE_CUDA_NUM_THREADS;
+  int max_blocks = prop.maxGridSize[0];
+  return num_blocks < max_blocks ? num_blocks : max_blocks;
```

Comment:
Is there is reason why you don't use `return std::min<int>(num_blocks, max_blocks);` ?